Add Colors screen for Design System

aditi-bhatia opened the pull request:
Part of issue https://github.com/wordpress-mobile/WordPress-Android/issues/19802

This PR adds the `Colors` screen to the design system section in the `Me` tab.

-----

## To Test:
1. Navigate to the `Me` tab. If you're on a debug build, you should see the `Design System` row.
2. Click `Foundation` -> `Colors`. 
3. You should see the new `Colors` screen and the title of the screen should update accordingly.
4. Go back to the `Me` tab and navigate to `App Settings` -> `Appearance` -> change the mode to dark or light mode. 
5. Go back to the `Colors` screen and you should see the colors updated (i.e Primary Foreground will be black in light mode and will be white in dark mode).

| Dark mode | Light mode |
|-----------|------------|
|   <img width="306" alt="Screenshot 2024-02-22 at 4 03 04 PM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/7199140/ae2e8c8e-feb7-4ebb-89e7-9cd975767c0a"> |  <img width="308" alt="Screenshot 2024-02-22 at 4 03 14 PM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/7199140/347440d9-958d-4709-ba5a-fca2b114f3c7">     |


-----

## Regression Notes

1. Potential unintended areas of impact

    - None

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    - None

7. What automated tests I added (or what prevented me from doing so)

    - None

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## Testing Checklist:

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

aditi-bhatia committed:
Add Colors screen

aditi-bhatia committed:
Add navigation to Colors screen + onClick

aditi-bhatia committed:
Merge branch 'trunk' of https://github.com/wordpress-mobile/WordPress-Android into feature/19802-designsystem-part4

aditi-bhatia committed:
Add color list

aditi-bhatia committed:
UI cleanup

dangermattic commented:
<!--
  0 Errors
  1 Warning: Class <code>ColorOption</code>...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ColorOption</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

aditi-bhatia committed:
Cleanup continued

aditi-bhatia committed:
Make rows smaller

ravishanker commented on WordPress/src/main/java/org/wordpress/android/designsystem/DesignSystemColorsScreen.kt:
```diff
@@ -0,0 +1,134 @@
+                        .size(45.dp)
+                        .clip(shape = RoundedCornerShape(5.dp, 5.dp, 5.dp, 5.dp))
+                        .background(color)
+                        .border(width = 2.dp,
```
Probably, don't need border here

aditi-bhatia replied:
I added a border because the background of each row is the `tertiaryBackground` color. Without the border, the row for that element blends in with the background, which wasn't looking so great IMO. I'm fine with keeping it or removing it (or making it thinner) though if we don't like how it looks, open to thoughts

ravishanker commented on WordPress/src/main/java/org/wordpress/android/designsystem/DesignSystemColorsScreen.kt:
```diff
@@ -0,0 +1,134 @@
+    Row (modifier = Modifier.padding(all = 3.dp)) {
+        Box (Modifier.background(MaterialTheme.colorScheme.tertiaryContainer)) {
+            Column(
+                horizontalAlignment = Alignment.CenterHorizontally,
```
Align this to left with the color square

aditi-bhatia replied:
Updated

ravishanker commented on WordPress/src/main/java/org/wordpress/android/designsystem/DesignSystemColorsScreen.kt:
```diff
@@ -0,0 +1,134 @@
+@Composable
+fun ColorCard (colorName: String, color: Color) {
+    Row (modifier = Modifier.padding(all = 3.dp)) {
+        Box (Modifier.background(MaterialTheme.colorScheme.tertiaryContainer)) {
```
Also, no background required here

aditi-bhatia replied:
Removed

ravishanker commented:
Here're screenshots of iOS that has colors separated into foreground, and background

| Dark | Light |
|--------|--------|
| ![IMG_2339](https://github.com/wordpress-mobile/WordPress-Android/assets/990349/226329ca-0149-4c79-9a11-6e7b7988e8b9) | ![IMG_617ECD9015E5-1](https://github.com/wordpress-mobile/WordPress-Android/assets/990349/7603989b-37cb-4972-89aa-f2d92ea30dfa)|

aditi-bhatia committed:
Remove background color

aditi-bhatia commented:
Hi Ravi, thank you for your suggestions. I removed the background, separated the foreground/background colors, and aligned the text. I've kept the color icon border for now (see comment above) but made it thinner.  I wasn't sure how closely we wanted to follow the iOS designs since this is just an internal tool for devs, but I did notice iOS has a background color on the elements as well. What are your thoughts?
<img width="408" alt="Screenshot 2024-02-23 at 2 48 54 PM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/7199140/6557a059-837e-4da4-b2f5-47a22eaa3825">

aditi-bhatia committed:
Separate colors into sections, create titles

aditi-bhatia committed:
Update preview

ravishanker commented:
> Hi Ravi, thank you for your suggestions. I removed the background, separated the foreground/background colors, and aligned the text. I've kept the color icon border for now (see comment above) but made it thinner. I wasn't sure how closely we wanted to follow the iOS designs since this is just an internal tool for devs, but I did notice iOS has a background color on the elements as well. What are your thoughts? 

👋 Aditi - Thanks for making the updates.
- [ ] Besides Foreground, Background items, there's also the third item **Divider** (see the light background iOS screenshot above)
- [x] Please remove the border, and 
- [x] add a list item divider, then it looks much better (see below)
- [x] Navigation title at the top should match the screen for e.g. Colors (see iOS screenshot above).  Useful [article on this](https://medium.com/bumble-tech/refining-compose-api-for-design-systems-d652e2c2eac3) 

| Light | Dark |
|--------|--------|
| ![Screenshot_20240226_121823](https://github.com/wordpress-mobile/WordPress-Android/assets/990349/32528ff0-5f57-4ab3-88a1-ad3d548613d2) | ![Screenshot_20240226_121832](https://github.com/wordpress-mobile/WordPress-Android/assets/990349/a476239e-b94a-43d2-b3bf-8cf6bc4463f8) |

ravishanker commented on WordPress/src/main/java/org/wordpress/android/designsystem/DesignSystemColorsScreen.kt:
```diff
@@ -0,0 +1,120 @@
+                    .size(45.dp)
+                    .clip(shape = RoundedCornerShape(5.dp, 5.dp, 5.dp, 5.dp))
+                    .background(color)
+                    .border(width = 1.dp,
```
Remove border here

aditi-bhatia replied:
Updated

ravishanker commented on WordPress/src/main/java/org/wordpress/android/designsystem/DesignSystemColorsScreen.kt:
```diff
@@ -0,0 +1,120 @@
+                color = MaterialTheme.colorScheme.secondary
+            )
+        }
+    }
```
Add a Divider(modifier = Modifier.padding(start = 10.dp, end = 10.dp)) after this brace }

aditi-bhatia replied:
Thank you, updated

ravishanker replied:
Did you forget to push this change?

aditi-bhatia replied:
Hey Ravi, thanks for checking - I'm still working on the other changes required for the navigation title. I'll tag you / re-request a review when this PR is updated and ready to be looked at again.

aditi-bhatia committed:
Remove border from color icons

aditi-bhatia committed:
Merge branch 'trunk' of https://github.com/wordpress-mobile/WordPress-Android into feature/19802-designsystem-part4

aditi-bhatia committed:
Add divider

aditi-bhatia committed:
Update screen title based on route navigation

aditi-bhatia committed:
Detekt

aditi-bhatia commented:
Hey Ravi, thanks for your input. I've updated the UI to remove the border and add the divider lines + also updated the navigation title for the screens. This is what it looks like now: 
<img width="436" alt="Screenshot 2024-02-27 at 12 39 28 AM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/7199140/966833c4-e313-475c-8bf8-118741b09d1a">


I took a look at the [tutorial](https://medium.com/bumble-tech/refining-compose-api-for-design-systems-d652e2c2eac3) you sent above which relaxes the API for the top bar. In this case, only the title is changing from screen to screen, so I didn't think it would be necessary to build out that abstraction  - but I'm open to discussing it further if you think it would be wise to implement that here. Let me know what your thoughts are and appreciate any other pointers/tips 👍

ravishanker commented on WordPress/src/main/java/org/wordpress/android/designsystem/DesignSystemColorsScreen.kt:
```diff
@@ -0,0 +1,118 @@
+                ColorOption("Tertiary", MaterialTheme.colorScheme.tertiaryContainer),
+                ColorOption("Brand", MaterialTheme.colorScheme.brandContainer),
+                ColorOption("WP", MaterialTheme.colorScheme.wpContainer),
+            )
```
Add the Divider option here before merging

aditi-bhatia replied:
@ravishanker Do you mean to make the last divider thicker? 
<img width="274" alt="Screenshot 2024-02-27 at 3 07 22 PM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/7199140/381c1b9e-0671-437c-9a79-3a7dc1439927">

ravishanker replied:
No, There're three categories Foreground, Background, and **Divider** [on iOS](https://github.com/wordpress-mobile/WordPress-Android/pull/20259#issuecomment-1960672133)

aditi-bhatia replied:
Yes, I noticed that too but I don't see a definition for that in the [Figma](https://www.figma.com/file/LyzJJa6ANKToUswWmRrZee/%F0%9F%A7%B1-Jetpack-App-Design-System-iOS?type=design&node-id=102-186&mode=design&t=HR4dAG6naRMBnhmd-0) which is why I skipped that category. 

Compared to the Figma, the iOS implementation also has an extra color for `Success` and `Quartenary Foreground` (in addition to `Quartenary Background`), and is missing the `WP Foreground`, `WP Background` colors that are in the Figma.

I was using the Figma definitions as the source of truth, but let me know if I should be following the iOS implementation instead. Also wanted to mention that the colors on Android will likely need to be updated as Chris [mentioned](https://a8c.slack.com/archives/C04TXQU42HM/p1708599421092949) both platforms will have different color definitions.

ravishanker replied:
For now, let's just match the iOS.

aditi-bhatia replied:
Hmm, I think it might make sense to check with Chris and Alp before making changes to match iOS instead of Figma. What do you think? 

I started looking at the values on iOS and noticed that a few values seem incorrect/incomplete (first screenshot) or are missing entirely (second screenshot). This is in addition to differing values from the Figma that I mentioned in the comment above, so there seems to be a few bugs and/or some confusion here. 

Screenshots from iOS:


<img width="353" alt="Screenshot 2024-02-27 at 5 50 29 PM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/7199140/71a0d4f8-424d-4583-86c0-c8cb7fe98a22"> 

<img width="353" alt="Screenshot 2024-02-27 at 5 50 41 PM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/7199140/4f8e228c-c932-4053-a23a-dce1ef207f65">

ravishanker reviewed (approved):
LGTM 👍

aditi-bhatia committed:
Fix quartenary color naming to quartenaryContainer since it's a background color (not foreground)

aditi-bhatia committed:
Typo in quaternary

aditi-bhatia commented:
Based on our conversation yesterday, I'll go ahead and merge this PR. I've created a separate [ticket](https://github.com/wordpress-mobile/WordPress-Android/issues/20345) to update the colors once we have the updated values from Design.
